### ISSUE TITLE
Demo cleanup/archival automation

### ISSUE THREAD

**Issue by vttranlina** [linagora/tmail-backend#828]
- record video 
- demo on super thursday

**Comment by quantranhong1999** [comment]
## Demo INBOX archival

https://github.com/linagora/tmail-backend/assets/55171818/5e4d4dbd-fa68-47db-be03-a6104868184e

### Bug fixes / issues
- https://github.com/linagora/tmail-backend/pull/845
- https://github.com/linagora/james-project/issues/4885

### Task runtime on preprod
#### Only user29 with 8K messages to be archived
<details>
 <summary>8442 messages archived over 5 minutes => ~ 14msg/second</summary>
TaskId: 07f94f52-634e-4de6-932e-dba10c88e56a

```json
{
    "additionalInformation": {
        "type": "InboxArchivalTask",
        "timestamp": "2023-10-10T05:45:29.744999870Z",
        "successfulUsersCount": 1,
        "failedUsersCount": 0,
        "failedUsers": [],
        "archivedMessageCount": 8442,
        "errorMessageCount": 0
    },
    "type": "InboxArchivalTask",
    "submittedFrom": "tmail-imap-smtp-74cb98875-rn86g",
    "executedOn": "tmail-jmap-58675946f5-6jrlw",
    "cancelledFrom": null,
    "submitDate": "2023-10-10T05:40:33.253+0000",
    "status": "completed",
    "taskId": "07f94f52-634e-4de6-932e-dba10c88e56a",
    "startedDate": "2023-10-10T05:40:33.392+0000",
    "completedDate": "2023-10-10T05:45:29.786+0000"
}
```
</details>

#### Only user29 with no messages to be archived ~ Task with no actually load

<details>
 <summary>INBOX archival with no load -> 25 second runtime</summary>
TaskId: 1204db78-d098-41fc-948b-5f9d8bde1b13

```json
{
    "additionalInformation": {
        "type": "InboxArchivalTask",
        "timestamp": "2023-10-10T07:03:36.807395441Z",
        "successfulUsersCount": 1,
        "failedUsersCount": 0,
        "failedUsers": [],
        "archivedMessageCount": 0,
        "errorMessageCount": 0
    },
    "type": "InboxArchivalTask",
    "submittedFrom": "tmail-imap-smtp-74cb98875-rn86g",
    "executedOn": "tmail-jmap-58675946f5-6jrlw",
    "cancelledFrom": null,
    "submitDate": "2023-10-10T07:03:09.138+0000",
    "startedDate": "2023-10-10T07:03:09.193+0000",
    "status": "completed",
    "taskId": "1204db78-d098-41fc-948b-5f9d8bde1b13",
    "completedDate": "2023-10-10T07:03:36.832+0000"
}
```
</details>

What I can get from the INBOX archival task runtime:
- Should not overload TMail while handling other user requests.
- Runtime could be a bit long upon the first run with many messages to archive, but should be quick the next times.

**Comment by chibenwa** [comment]
Stupid question: is tung fix on OpenSearch messageSearch index making it any faster?
Can we schedule several users in parallel?

Please confirm that defzult behaviour for all automations is "disabled"

**Comment by chibenwa** [comment]
By assuming uid order matches sentDate order, we maybe could optimize a lot the "no message to move" use case and make it waaaaay faster, no?

**Comment by vttranlina** [comment]
> Stupid question: is tung fix on OpenSearch messageSearch index making it any faster? Can we schedule several users in parallel?
> 
> Please confirm that defzult behaviour for all automations is "disabled"

It just merged this morning, It is not available in tmail master branch,  
// In the past, I benchmarked on sandbox by my branch

**Comment by vttranlina** [comment]
> By assuming uid order matches sentDate order, we maybe could optimize a lot the "no message to move" use case and make it waaaaay faster, no?

"order" here is "order" in database?
Not sure the `order` in db will be faster than "fetch all & filter" in code

**Comment by quantranhong1999** [comment]
## Trash cleanup Demo

https://github.com/linagora/tmail-backend/assets/55171818/ca5bdc85-a046-452e-8d9d-24026d5b38cb

### Task runtime on preprod
#### Only user30, usersPerSecond=1

Too long TBH, it takes hours and I did not have the patience to wait for it to finish...

#### Only user30 with 1 message in Trash, usersPerSecond=50

<details>
 <summary>It takes 10 minutes to finish the tasks</summary>
TaskId: b2bc74ea-498f-486e-8c9f-6fd09a53aa10

```json
{
    "additionalInformation": {
        "type": "cleanup-trash",
        "processedUsersCount": 31384,
        "deletedMessagesCount": 1,
        "failedUsers": [],
        "runningOptions": {
            "usersPerSecond": 50
        },
        "timestamp": "2023-10-10T09:27:05.210684222Z"
    },
    "type": "cleanup-trash",
    "submittedFrom": "tmail-jmap-58675946f5-6jrlw",
    "executedOn": "tmail-jmap-58675946f5-6jrlw",
    "cancelledFrom": null,
    "submitDate": "2023-10-10T09:16:36.060+0000",
    "status": "completed",
    "taskId": "b2bc74ea-498f-486e-8c9f-6fd09a53aa10",
    "startedDate": "2023-10-10T09:16:36.111+0000",
    "completedDate": "2023-10-10T09:27:05.241+0000"
}
```
</details>

What I can get from the Trash cleanup task:
- It is slow generally even with no load (no messages to deleted). My guess is that the throttler is what slows the task down.
- Should run with high enough usersPerSecond.

**Comment by vttranlina** [comment]
> It is slow generally even with no load (no messages to deleted). My guess is that the throttler is what slows the task down.

users = 31384
usersPerSecond = 50
=> Spent time = 31384/50 = ~10 minutes, 
It is normal

**Comment by quantranhong1999** [comment]
## Spam cleanup Demo

https://github.com/linagora/tmail-backend/assets/55171818/b28cfc4b-2ae4-4223-bf7c-9790dbeb78be

Basically the same as the Trash cleanup.

**Comment by quantranhong1999** [comment]
PR to put higher `usersPerSecond` for preprod cron jobs: https://ci.linagora.com/linagora/lrs/saas/deployments/tmail/emaily-staging-apps/-/merge_requests/16

**Comment by Arsnael** [comment]
50 users per sercond? Isn't that a bit much?

**Comment by quantranhong1999** [comment]
> 50 users per second? Isn't that a bit much?

How much do you suggest then?

**Comment by Arsnael** [comment]
No idea... I mean did you try on preprod for example?

My guess is: 
- should absolutely not run it at more than 1 user per second the first time, as there would be a lot of messages to treat
- after the first time though maybe 50 is ok IDK, as the number of messages concerned should stay relatively low if you run it every day. If you tell me you can run it now on preprod with 50 and without issues, I'm fine then :)

**Comment by quantranhong1999** [comment]
> I mean did you try on preprod for example?

As I commented above https://github.com/linagora/tmail-backend/issues/828#issuecomment-1754854763
For the case: only 1 user (over 33k users) enabled cleanup, and just has 1 message to clean up:
- usersPerSecond=1 -> The task could take a day to finish. Even most users do not need to clean up.
- usersPerSecond=50 -> The task takes about 10 minutes to finish. Preprod was fine then. But with just 1 user enabled so I am not sure.

**Comment by Arsnael** [comment]
If you try to enable it on the first time for a 100 users (which is not crazy seeing there is 20k users on the platform) and you put 50 users per second, I wouldnt be surprised it goes boom tbh :)

**Comment by quantranhong1999** [comment]
I run again tasks with 30 users enabled (user 29 -> user 58).

### Clean up Trash/Spam
usersPerSecond = **20**

Preprod is good (no error logs). Tasks take 26 minutes to finish. Acceptable runtime IMO.

```json
{
    "additionalInformation": {
        "type": "cleanup-spam",
        "processedUsersCount": 30,
        "deletedMessagesCount": 555,
        "failedUsers": [],
        "runningOptions": {
            "usersPerSecond": 20
        },
        "timestamp": "2023-10-11T07:54:53.154933936Z"
    },
    "type": "cleanup-spam",
    "submittedFrom": "tmail-jmap-7c47449bb8-kz7tp",
    "executedOn": "tmail-jmap-7c47449bb8-kz7tp",
    "cancelledFrom": null,
    "submitDate": "2023-10-11T07:28:40.911+0000",
    "status": "completed",
    "taskId": "a1b24ec9-7d8a-40c4-99a2-76317ab6ff61",
    "completedDate": "2023-10-11T07:54:53.185+0000",
    "startedDate": "2023-10-11T07:28:41.035+0000"
}
```

IMO normally there are not many messages in Spam/Trash (compared to INBOX), so we can use a bit of high `usersPerSecond`. BTW `usersPerSecond` should be scaled upon deployment size, as yes agree with @Arsnael we should be careful upon first runs.

### Inbox archival

At first it went boom with `CassandraTimeoutException` actually, so I reduce the task concurrency: https://github.com/linagora/tmail-backend/pull/846

After the fix, preprod is fine with the INBOX archival task.

```json
{
    "additionalInformation": {
        "type": "InboxArchivalTask",
        "timestamp": "2023-10-11T06:40:48.045723776Z",
        "successfulUsersCount": 30,
        "failedUsersCount": 0,
        "failedUsers": [],
        "archivedMessageCount": 83833,
        "errorMessageCount": 0
    },
    "type": "InboxArchivalTask",
    "submittedFrom": "tmail-imap-smtp-5997867cb9-sjqvt",
    "executedOn": "tmail-jmap-7c47449bb8-kz7tp",
    "cancelledFrom": null,
    "submitDate": "2023-10-11T06:16:31.774+0000",
    "status": "completed",
    "taskId": "a0b4831d-fded-4c24-9553-93531296c002",
    "completedDate": "2023-10-11T06:40:48.067+0000",
    "startedDate": "2023-10-11T06:16:32.719+0000"
}
```

**Comment by quantranhong1999** [comment]
> If you try to enable it on the first time for a 100 users (which is not crazy seeing there is 20k users on the platform) and you put 50 users per second, I wouldnt be surprised it goes boom tbh :)

I reduced it to 20 usersPerSecond. And tests on preprod look good to me (just 30 users enable BTW). Is it better to you?

**Comment by quantranhong1999** [comment]
I think listing all users' INBOX messages is heavy.

How about using OpenSearch to search messages before a date? That would avoid listing and iterate through all INBOX messages.
And rely on `internalDate` (I wished to be able to use `saveDate` but not always the case for search). Relying on internalDate for INBOX is acceptable IMO, while it is not for other mailboxes that rely more on moved messages like TRASH/SPAM.

WDYT?

**Comment by vttranlina** [comment]
> How about using OpenSearch to search messages before a date?

IMO, this feature is "task", 
fast is not a higher priority than consistency for this case. 
we can decrease the userPerSecond to decrease the pressure on Cassandra in prod.

**Comment by quantranhong1999** [comment]
> fast is not a higher priority than consistency for this case.

Actually, I do not mean to prioritize the speed of the task, but to depress the pressure that the task put on TMail and Cassandra. If a user has e.g. a million messages in his INBOX to fetch, that would be a lot of pressure.

**Comment by vttranlina** [comment]
The optimized performance will discussed/continued on another ticket. 
The demo is done
